docs/dev/testing.md

```diff
@@ -90,32 +99,162 @@ reject file:
+steps. While this speeds up execution, sharing servers complicates debugging
+if a test fails.
+
+Specifically, you should avoid leaving global artifacts in your test, even
+if it fails. Typically, you could use a built-in `keyspace()` fixture
+to create a randomly named keyspace.
+
+At start and end of each test, `test.py` performs a number of sanity checks
+of the used server:
+- it should be up and running,
+- it should not contain non-system keyspaces.
+
+### Debugging a pytest.
+
+To have a full picture for a failing pytest it is necessary to identify
+the server which was used to run it and the relevant framgnet in the server
+log. Thankfuly, `test.py` maintains this link through relevant log
+messages and preserves Scylla output on test failure.
+
+A typical debugging journey should start with looking at `test.py.log` in
+`testlog` where, for each test it runs, `test.py` prints all relevant paths
+to server log and pytest output.
+
+To extend `test.py` logging, you can use the standard 'logging' module API.
+Individual pytests are programmed to not gobble stdout, so you can can also
+add prints to pytests, and they will end up in the test' log.
+
+For example, imaigine `cql-pytest/test_null.py` fails. The relevant lines
+in test.py.log will be:
+
+```
+21:53:04.789 INFO> Created cluster {127.101.161.1}
+21:53:04.790 INFO> Leasing Scylla cluster {127.101.161.1} for test test_null.1
+21:53:04.790 INFO> Starting test test_null.1: pytest --host=127.101.161.1 -s ...test/cql-pytest/test_null.py
+21:53:05.533 INFO> Test test_null.1 failed
+```
+To find out the working directory of instance 127.101.161.1 search
+for its initialization message in `test.py.log`:
+
+```
+10:05:51.722 INFO> installing Scylla server in /opt/local/work/scylla/scylla/testlog/dev/scylla-1...
+10:05:51.722 INFO> starting server at host 127.159.235.1 in scylla-1...
+10:05:52.688 INFO> started server at host 127.159.235.1 in scylla-1, pid 2165602
+```
+
+Next, we can take a look at the server log, which is at
+`/opt/local/work/scylla/scylla/testlog/dev/scylla-1.log`:
+
+The log contains special markers, written at test start and end:
+
+```
+INFO  2022-08-18 10:05:52,598 [shard 0] schema_tables - Schema version changed to 8b5e9c73-7c1c-3b28-8c31-c1359210c484
+------ Starting test test_null.1 ------
+...
+INFO  2022-08-18 10:05:53,297 [shard 0] schema_tables - Dropping keyspace cql_test_1660806353124
+INFO  2022-08-18 10:05:53,304 [shard 0] schema_tables - Schema version changed to 8b5e9c73-7c1c-3b28-8c31-c1359210c484
+------ Ending test test_null.1 ------
+```
+
+Most often there are no errors in Scylla log, so next we inspect
+the test' log, which is next to the server's at
+`/opt/local/scylla/scylla/testlog/dev/test_null.1.log`:
+```
+        cql.execute(f"INSERT INTO {table1} (p,c) VALUES ('{p}', '3')")
+>       assert False
+E       assert False
+```
+
+What does number 1 mean in the log file name? Since test.py can
```

Comment:
`test.py` in single quotes for consistency?

Reply:
fixed